File: src/symphonic_cipher/scbe_aethermoore/game/companion.py
"""
Seal Entity — Companion System (Python reference).

Mirrors src/game/companion.ts.
Companions are 21D canonical state vectors with behavioral evolution.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import List, Optional

from .types import (
    BondType,
    CanonicalState,
    DisciplineTrait,
    EggType,
    EmotionalState,
    EvolutionStage,
    EVOLUTION_THRESHOLDS,
    FormationRole,
    OVER_EVOLUTION_THRESHOLD,
    TongueCode,
    TongueVector,
    TONGUE_CODES,
    default_canonical_state,
    tongue_norm,
)


@dataclass
class DerivedCombatStats:
    speed: float
    insight: float
    perception: float
    risk_tolerance: float
    entropy_affinity: float
    resilience: float
    proof_power: float
    authority: float


def derive_combat_stats(state: CanonicalState) -> DerivedCombatStats:
    """Derive combat stats from 21D canonical state. Never set directly."""
    clamp = lambda x: max(0.0, min(100.0, x * 100))
    return DerivedCombatStats(
        speed=clamp(state.flux),
        insight=clamp(state.coherence_s),
        perception=clamp(state.coherence_tri),
        risk_tolerance=clamp(state.risk),
        entropy_affinity=clamp(state.entropy_rate),
        resilience=clamp(state.stabilization),
        proof_power=clamp(state.radius),
        authority=clamp(state.harmonic_energy),
    )


@dataclass
class Companion:
    """A Seal Entity — companion built on 21D canonical state."""

    id: str
    species_id: str
    name: str
    state: CanonicalState
    seal_integrity: float = 100.0
    max_seal_integrity: float = 100.0
    drift_level: float = 0.0
    bond_level: int = 1
    bond_xp: int = 0
    formation_role: FormationRole = "storm"
    discipline_trait: DisciplineTrait = "collaborative"
    emotional_state: EmotionalState = "content"
    evolution_stage: EvolutionStage = "spark"
    evolution_line: List[str] = field(default_factory=list)
    egg_origin: EggType = "mono_KO"
    bond_type: BondType = "amplifier"
    scar_count: int = 0
    hollow_exposure: int = 0

    @property
    def derived_stats(self) -> DerivedCombatStats:
        return derive_combat_stats(self.state)


def create_companion(
    id: str,
    species_id: str,
    name: str,
    egg_type: EggType,
    bond_type: BondType,
    initial_tongue: TongueVector,
) -> Companion:
    """Create a new companion from a hatched egg."""
    state = CanonicalState(tongue_position=initial_tongue, radius=0.1)
    return Companion(
        id=id,
        species_id=species_id,
        name=name,
        state=state,
        evolution_line=[species_id],
        egg_origin=egg_type,
        bond_type=bond_type,
    )


def apply_tongue_experience(comp: Companion, tongue: TongueCode, amount: float) -> None:
    """Shift tongue position toward the given tongue. A2: Unitarity preserved."""
    idx = TONGUE_CODES.index(tongue)
    pos = list(comp.state.tongue_position)
    old_norm = tongue_norm(tuple(pos))  # type: ignore[arg-type]
    pos[idx] = min(1.0, pos[idx] + amount * 0.1)
    new_norm = tongue_norm(tuple(pos))  # type: ignore[arg-type]

    if new_norm > 0 and old_norm > 0:
        target_norm = min(1.0, old_norm + amount * 0.02)
        scale = target_norm / new_norm
        pos = [x * scale for x in pos]

    comp.state = CanonicalState(
        tongue_position=tuple(pos),  # type: ignore[arg-type]
        phase_angles=comp.state.phase_angles,
        flux=comp.state.flux,
        coherence_s=comp.state.coherence_s,
        coherence_bi=comp.state.coherence_bi,
        coherence_tri=comp.state.coherence_tri,
        risk=comp.state.risk,
        entropy_rate=comp.state.entropy_rate,
        stabilization=comp.state.stabilization,
        radius=comp.state.radius,
        harmonic_energy=comp.state.harmonic_energy,
    )


def apply_combat_result(comp: Companion, won: bool, difficulty: int) -> None:
    """Apply combat result to companion state."""
    diff_norm = difficulty / 10.0

    if won:
        radius_gain = 0.01 + diff_norm * 0.02
        coh_gain = 0.005 + diff_norm * 0.01
        comp.state = CanonicalState(
            tongue_position=comp.state.tongue_position,
            phase_angles=comp.state.phase_angles,
            flux=comp.state.flux,
            coherence_s=min(1.0, comp.state.coherence_s + coh_gain),
            coherence_bi=comp.state.coherence_bi,
            coherence_tri=comp.state.coherence_tri,
            risk=comp.state.risk,
            entropy_rate=comp.state.entropy_rate,
            stabilization=min(1.0, comp.state.stabilization + coh_gain * 0.5),
            radius=min(1.0, comp.state.radius + radius_gain),
            harmonic_energy=min(1.0, comp.state.harmonic_energy + 0.005),
        )
        comp.bond_xp += 5 + difficulty
    else:
        comp.scar_count += 1
        entropy_gain = diff_norm * 0.03
        comp.state = CanonicalState(
            tongue_position=comp.state.tongue_position,
            phase_angles=comp.state.phase_angles,
            flux=comp.state.flux,
            coherence_s=comp.state.coherence_s,
            coherence_bi=comp.state.coherence_bi,
            coherence_tri=comp.state.coherence_tri,
            risk=min(1.0, comp.state.risk + 0.01),
            entropy_rate=min(1.0, comp.state.entropy_rate + entropy_gain),
            stabilization=comp.state.stabilization,
            radius=comp.state.radius,
            harmonic_energy=comp.state.harmonic_energy,
        )
        comp.seal_integrity = max(0, comp.seal_integrity - (10 + difficulty * 3))
        comp.bond_xp += 2

    # Bond level up
    xp_for_next = comp.bond_level * 10
    if comp.bond_xp >= xp_for_next and comp.bond_level < 10:
        comp.bond_level += 1
        comp.bond_xp -= xp_for_next


def current_evolution_stage(radius: float) -> EvolutionStage:
    """Determine evolution stage from radius."""
    if radius >= EVOLUTION_THRESHOLDS["transcendent"]:
        return "transcendent"
    if radius >= EVOLUTION_THRESHOLDS["apex"]:
        return "apex"
    if radius >= EVOLUTION_THRESHOLDS["prime"]:
        return "prime"
    if radius >= EVOLUTION_THRESHOLDS["form"]:
        return "form"
    return "spark"


def is_over_evolved(comp: Companion) -> bool:
    """Check if companion is unstable (radius > 0.95)."""
    return comp.state.radius > OVER_EVOLUTION_THRESHOLD

File: src/game/companion.ts
/**
 * @file companion.ts
 * @module game/companion
 * @layer Layer 5, Layer 12, Layer 13
 * @component Seal Entity — Companion System
 *
 * Companions are 21D canonical state vectors with behavioral evolution,
 * symbiotic graph topology, and fleet formation roles.
 * A2: Unitarity — state transitions preserve norm within tolerance.
 * A4: Symmetry — Hodge dual pairs receive 30% bond bonus.
 */

import {
  CanonicalState,
  defaultCanonicalState,
  stateToArray,
  DisciplineTrait,
  EmotionalState,
  FormationRole,
  EvolutionStage,
  EVOLUTION_THRESHOLDS,
  OVER_EVOLUTION_THRESHOLD,
  TongueCode,
  TongueVector,
  TONGUE_CODES,
  dominantTongue,
  tongueDistance,
  tongueNorm,
  EggType,
  BondType,
} from './types.js';

// ---------------------------------------------------------------------------
//  Companion (Seal Entity)
// ---------------------------------------------------------------------------

export interface Companion {
  /** Unique identifier */
  readonly id: string;

  /** Display name */
  name: string;

  /** Species/lineage identifier */
  readonly speciesId: string;

  /** The 21D canonical state — THIS IS the companion */
  state: CanonicalState;

  /** Current seal integrity (HP equivalent, 0-100) */
  sealIntegrity: number;

  /** Maximum seal integrity */
  maxSealIntegrity: number;

  /** Current drift level (distance from stable alignment) */
  driftLevel: number;

  /** Bond level with player (0-10) */
  bondLevel: number;

  /** Bond experience points toward next level */
  bondXP: number;

  /** Current formation role */
  formationRole: FormationRole;

  /** Discipline trait (develops from usage patterns) */
  disciplineTrait: DisciplineTrait;

  /** Emotional state (affects performance + evolution) */
  emotionalState: EmotionalState;

  /** Current evolution stage */
  evolutionStage: EvolutionStage;

  /** Evolution line (tracks the path taken) */
  evolutionLine: string[];

  /** Egg type this companion hatched from */
  eggOrigin: EggType;

  /** Bond type from egg origin */
  bondType: BondType;

  /** Scar count (losses, failed dungeons) — influences evolution */
  scarCount: number;

  /** Hollow exposure count (contact with 7th Tongue) */
  hollowExposure: number;

  /** Combat stats (derived from canonical state) */
  readonly derivedStats: DerivedCombatStats;
}

/** Stats derived from the 21D canonical state — never set directly */
export interface DerivedCombatStats {
  readonly speed: number;
  readonly insight: number;
  readonly perception: number;
  readonly riskTolerance: number;
  readonly entropyAffinity: number;
  readonly resilience: number;
  readonly proofPower: number;
  readonly authority: number;
}

// ---------------------------------------------------------------------------
//  Stat Derivation (canonical state → combat stats)
// ---------------------------------------------------------------------------

/**
 * Derive combat stats from the 21D canonical state.
 * These are COMPUTED, never assigned directly.
 */
export function deriveCombatStats(state: CanonicalState): DerivedCombatStats {
  return {
    speed: Math.max(0, Math.min(100, state.flux * 100)),
    insight: Math.max(0, Math.min(100, state.coherence_s * 100)),
    perception: Math.max(0, Math.min(100, state.coherence_tri * 100)),
    riskTolerance: Math.max(0, Math.min(100, state.risk * 100)),
    entropyAffinity: Math.max(0, Math.min(100, state.entropy_rate * 100)),
    resilience: Math.max(0, Math.min(100, state.stabilization * 100)),
    proofPower: Math.max(0, Math.min(100, state.radius * 100)),
    authority: Math.max(0, Math.min(100, state.harmonic_energy * 100)),
  };
}

// ---------------------------------------------------------------------------
//  Companion Factory
// ---------------------------------------------------------------------------

/**
 * Create a new companion from a hatched egg.
 *
 * @param id - Unique identifier
 * @param speciesId - Species lineage
 * @param name - Display name
 * @param eggType - Egg this hatched from
 * @param bondType - Bond type determined by egg
 * @param initialTongue - Starting tongue position
 */
export function createCompanion(
  id: string,
  speciesId: string,
  name: string,
  eggType: EggType,
  bondType: BondType,
  initialTongue: TongueVector
): Companion {
  const state: CanonicalState = {
    ...defaultCanonicalState(),
    tonguePosition: initialTongue,
    radius: 0.1, // start at spark stage
  };

  return {
    id,
    speciesId,
    name,
    state,
    sealIntegrity: 100,
    maxSealIntegrity: 100,
    driftLevel: 0,
    bondLevel: 1,
    bondXP: 0,
    formationRole: 'storm',
    disciplineTrait: 'collaborative',
    emotionalState: 'content',
    evolutionStage: 'spark',
    evolutionLine: [speciesId],
    eggOrigin: eggType,
    bondType,
    scarCount: 0,
    hollowExposure: 0,
    get derivedStats() {
      return deriveCombatStats(this.state);
    },
  };
}

// ---------------------------------------------------------------------------
//  State Transitions
// ---------------------------------------------------------------------------

/**
 * Apply tongue experience to a companion's state.
 * Shifts tongue position toward the given tongue.
 * A2: Unitarity — magnitude preserved within ε.
 *
 * @param comp - Companion to modify (mutated in place)
 * @param tongue - Which tongue gained experience
 * @param amount - Experience magnitude (0-1)
 */
export function applyTongueExperience(comp: Companion, tongue: TongueCode, amount: number): void {
  const idx = TONGUE_CODES.indexOf(tongue);
  const pos = [...comp.state.tonguePosition] as TongueVector;

  // Shift toward this tongue
  const oldNorm = tongueNorm(pos);
  pos[idx] = Math.min(1.0, pos[idx] + amount * 0.1);

  // A2: Unitarity — renormalize if we grew too much
  const newNorm = tongueNorm(pos);
  if (newNorm > 0 && oldNorm > 0) {
    const targetNorm = Math.min(1.0, oldNorm + amount * 0.02);
    const scale = targetNorm / newNorm;
    for (let i = 0; i < 6; i++) pos[i] *= scale;
  }

  comp.state = { ...comp.state, tonguePosition: pos };
}

/**
 * Apply combat result to companion state.
 *
 * @param comp - Companion to modify
 * @param won - Whether the encounter was won
 * @param difficulty - Problem difficulty (1-10)
 */
export function applyCombatResult(comp: Companion, won: boolean, difficulty: number): void {
  const diffNorm = difficulty / 10;

  if (won) {
    // Grow radius (proof power) — faster for harder problems
    const radiusGain = 0.01 + diffNorm * 0.02;
    const newRadius = Math.min(1.0, comp.state.radius + radiusGain);

    // Improve coherence
    const cohGain = 0.005 + diffNorm * 0.01;

    comp.state = {
      ...comp.state,
      radius: newRadius,
      coherence_s: Math.min(1.0, comp.state.coherence_s + cohGain),
      stabilization: Math.min(1.0, comp.state.stabilization + cohGain * 0.5),
      harmonic_energy: Math.min(1.0, comp.state.harmonic_energy + 0.005),
    };

    // Bond XP
    comp.bondXP += 5 + difficulty;
  } else {
    // Scar event
    comp.scarCount++;

    // Entropy exposure
    const entropyGain = diffNorm * 0.03;
    comp.state = {
      ...comp.state,
      entropy_rate: Math.min(1.0, comp.state.entropy_rate + entropyGain),
      risk: Math.min(1.0, comp.state.risk + 0.01),
    };

    // Seal integrity damage
    comp.sealIntegrity = Math.max(0, comp.sealIntegrity - (10 + difficulty * 3));

    // Bond XP (less, but still some for shared hardship)
    comp.bondXP += 2;
  }

  // Check bond level up (10 XP per level)
  const xpForNext = comp.bondLevel * 10;
  if (comp.bondXP >= xpForNext && comp.bondLevel < 10) {
    comp.bondLevel++;
    comp.bondXP -= xpForNext;
  }

  // Update emotional state
  comp.emotionalState = deriveEmotionalState(comp);
}

/**
 * Apply drift (chaos exposure) to a companion.
 * High drift = volatile: powerful but risky.
 */
export function applyDrift(comp: Companion, amount: number): void {
  comp.driftLevel = Math.min(1.0, Math.max(0, comp.driftLevel + amount));

  // Drift affects coherence negatively but can boost flux
  comp.state = {
    ...comp.state,
    coherence_s: Math.max(0, comp.state.coherence_s - amount * 0.1),
    flux: Math.min(1.0, comp.state.flux + amount * 0.05),
  };
}

/**
 * Rest/heal a companion — reduces drift, restores integrity.
 */
export function restCompanion(comp: Companion, amount: number): void {
  comp.sealIntegrity = Math.min(comp.maxSealIntegrity, comp.sealIntegrity + amount * 20);
  comp.driftLevel = Math.max(0, comp.driftLevel - amount * 0.1);
  comp.state = {
    ...comp.state,
    coherence_bi: Math.min(1.0, comp.state.coherence_bi + amount * 0.05),
  };
  comp.emotionalState = deriveEmotionalState(comp);
}

// ---------------------------------------------------------------------------
//  Emotional State Derivation
// ---------------------------------------------------------------------------

function deriveEmotionalState(comp: Companion): EmotionalState {
  if (comp.driftLevel > 0.8) return 'corrupted';
  if (comp.state.radius > OVER_EVOLUTION_THRESHOLD) return 'transcendent';
  if (comp.sealIntegrity < 20) return 'exhausted';
  if (comp.driftLevel > 0.5) return 'anxious';
  if (comp.state.radius > 0.7 && comp.bondLevel >= 5) return 'determined';
  if (comp.bondXP > 0 && comp.sealIntegrity > 80) return 'excited';
  return 'content';
}

// ---------------------------------------------------------------------------
//  Evolution Check
// ---------------------------------------------------------------------------

/**
 * Determine the current evolution stage from radius.
 */
export function currentEvolutionStage(radius: number): EvolutionStage {
  if (radius >= EVOLUTION_THRESHOLDS.transcendent) return 'transcendent';
  if (radius >= EVOLUTION_THRESHOLDS.apex) return 'apex';
  if (radius >= EVOLUTION_THRESHOLDS.prime) return 'prime';
  if (radius >= EVOLUTION_THRESHOLDS.form) return 'form';
  return 'spark';
}

/**
 * Check if a companion is ready to evolve.
 * Returns the new stage or null if no evolution available.
 */
export function checkEvolution(comp: Companion): EvolutionStage | null {
  const newStage = currentEvolutionStage(comp.state.radius);
  if (newStage !== comp.evolutionStage) {
    return newStage;
  }
  return null;
}

/**
 * Check if companion is over-evolved (unstable).
 * Over-evolved companions may disobey, mutate, or drift.
 */
export function isOverEvolved(comp: Companion): boolean {
  return comp.state.radius > OVER_EVOLUTION_THRESHOLD;
}

// ---------------------------------------------------------------------------
//  Discipline Trait Development
// ---------------------------------------------------------------------------

/**
 * Update discipline trait based on cumulative usage patterns.
 *
 * @param comp - Companion to update
 * @param verifyCount - Number of times used for verification
 * @param soloCount - Number of solo combat encounters
 * @param teamCount - Number of team/fleet encounters
 * @param riskCount - Number of high-risk actions taken
 */
export function updateDisciplineTrait(
  comp: Companion,
  verifyCount: number,
  soloCount: number,
  teamCount: number,
  riskCount: number
): void {
  const total = verifyCount + soloCount + teamCount + riskCount;
  if (total < 10) return; // need enough data

  const verifyRatio = verifyCount / total;
  const soloRatio = soloCount / total;
  const teamRatio = teamCount / total;
  const riskRatio = riskCount / total;

  if (verifyRatio > 0.4) comp.disciplineTrait = 'careful_verifier';
  else if (soloRatio > 0.4) comp.disciplineTrait = 'solo';
  else if (teamRatio > 0.4) comp.disciplineTrait = 'collaborative';
  else if (riskRatio > 0.4) comp.disciplineTrait = 'risk_tolerant';
  else if (comp.sealIntegrity > 90 && comp.state.stabilization > 0.7)
    comp.disciplineTrait = 'guardian';
  else comp.disciplineTrait = 'fast_heuristic';
}
